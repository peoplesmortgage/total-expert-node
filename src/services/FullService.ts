import { Service } from 'typedi';
import ServiceWrapper from './PartialService';
import { TotalExpertId, UpdateResponse } from '../types';

// as of v1 in 12/2020 - not all entities support all the CRUD actions through the API.
// If they do - the service can extend this class to get these operations for free,
// if not they should extend base class of ServiceWrapper

@Service()
class FullCrudTEService<T, E> extends ServiceWrapper<T, E> {
  constructor(urlSlug: string) {
    super(urlSlug);
  }

  async update(id: TotalExpertId, updateData: E): Promise<UpdateResponse> {
    const options = { method: 'PATCH', body: JSON.stringify(updateData) };
    const response = await this.auth.withAuthFetch(`${this.urlSlug}/${id}`, options);
    const data: UpdateResponse = await response.json();
    return data;
  }

  async delete(id: TotalExpertId): Promise<void> {
    await this.auth.withAuthFetch(`${this.urlSlug}/${id}`, { method: 'DELETE' });
  }
}

export default FullCrudTEService;
