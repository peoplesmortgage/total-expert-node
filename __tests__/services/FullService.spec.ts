import 'reflect-metadata';
import FullService from '../../src/services/FullService';
import mockAuthentication, { mockAPIReturnValue } from '../__fixtures__/mockAuthentication';

describe('FullService', () => {
  const mockAuth = mockAuthentication();

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('stores its constructor argument as the urlSlug protected property', () => {
    expect(new FullService<any, any>('/the-url-slug-value')).toMatchSnapshot();
  });

  it('exposes an update method that returns the response json of the api call', async () => {
    const returnJson = 'some return data';
    const mockResponseJson = mockAPIReturnValue(returnJson);
    const result = await new FullService('/test').update('test-id', {});

    expect.assertions(2);
    expect(mockResponseJson).toHaveBeenCalled();
    expect(result).toEqual(returnJson);
  });

  it('sets the http verb to PATCH when calling the update method', async () => {
    mockAPIReturnValue();
    await new FullService('/test').update('test-id', {});
    const { method } = mockAuth.withAuthFetch.mock.calls[0][1];
    expect(method).toEqual('PATCH');
  });

  it('sets the http verb to DELETE when hitting the API in the delete method', async () => {
    mockAPIReturnValue();
    await new FullService('/test').delete('test-id');
    const { method } = mockAuth.withAuthFetch.mock.calls[0][1];
    expect(method).toEqual('DELETE');
  });
});
