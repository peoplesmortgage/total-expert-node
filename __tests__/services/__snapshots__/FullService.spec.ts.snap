// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullService stores its constructor argument as the urlSlug protected property 1`] = `
FullCrudTEService {
  "auth": Object {
    "authenticate": [MockFunction],
    "setAccessToken": [MockFunction],
    "withAuthFetch": [MockFunction],
  },
  "urlSlug": "/the-url-slug-value",
}
`;
